### ISSUE TITLE
Sora no kiseki HD Block transfer is not correct

### ISSUE THREAD

**Issue by daniel229** [hrydgard/ppsspp#8335]
It works fine with sofware rendering.
The video rename jpg to mp4
1x
![ppssppwindows64v1 1 1-500-g2909580 2015-12-30 14-40-07-916](https://cloud.githubusercontent.com/assets/3481559/12047277/e84ab2e4-af03-11e5-845a-36c05d76adc5.jpg)

4x
![ppssppwindows64v1 1 1-500-g2909580 2015-12-30 14-46-52-880](https://cloud.githubusercontent.com/assets/3481559/12047302/3c5acacc-af04-11e5-978e-e079ad62f7b3.jpg)

sofware rendering
![ppssppwindows64v1 1 1-500-g2909580 2015-12-30 14-48-54-739](https://cloud.githubusercontent.com/assets/3481559/12047321/8968a85c-af04-11e5-8621-20b76868140e.jpg)

**Comment by unknownbrackets** [comment]
hrydgard had a very interesting theory: it could be this is being drawn with NEAREST, and drifting more each frame due to a download/texture loop or something.  Which sounds pretty likely, actually.

Since softgpu doesn't slide at all, it's probably not meant to slide at all.  If you force LINEAR for the texture filtering, does it stop happening?

-[Unknown]

**Comment by daniel229** [comment]
Force LINEAR,it looks like the 4x one.
Force NEAREST,it looks like the 1x one.
Still not like the software redering.

**Comment by unknownbrackets** [comment]
If you use the GE debugger, what is the vertex format (under Settings) it's using when it draws this?

I wonder if this is related to the texture coordinate doubling in some way.

-[Unknown]

**Comment by daniel229** [comment]
Is It this?
![04](https://cloud.githubusercontent.com/assets/3481559/12059739/f2422294-af9b-11e5-81a0-fb7788049a12.png)

The background picture also has a little offset.

**Comment by unknownbrackets** [comment]
That looks right.

Hmm.  So throughmode with float texcoords.  I guess we could manually round them maybe?  Actually the red box even looks slightly offset.  What do the vertices look like?

-[Unknown]

**Comment by daniel229** [comment]
Here
![10](https://cloud.githubusercontent.com/assets/3481559/12060260/a747d292-afa6-11e5-9d90-e13ef5349817.png)

**Comment by unknownbrackets** [comment]
Well, the XY and UV are both "wrong."  This really makes me think that the slide of the screen could be intentional.  I wonder how it looks on a PS3 (since it is a remaster, after all.)  It might also have the behavior we do at >= 2x.

Nearest in this case, I *think* (might want a hw test to verify) should predictably round the UVs downward, making the UVs above match the XY.  But I suppose the behavior of LINEAR is "correct."

Sorry for asking for another screenshot, but how about the texture tab?  I'm wondering what the filtering setting is set to exactly - if it's linear always or just linear when magnifying.

-[Unknown]

**Comment by daniel229** [comment]
It's nearest.
![11](https://cloud.githubusercontent.com/assets/3481559/12060367/d715aad8-afa8-11e5-8442-b846a6ea856b.png)

**Comment by daniel229** [comment]
I found the video for SC play on PS3 the screen is like the software.
https://youtu.be/n94plorFBXE?t=18m25s

**Comment by unknownbrackets** [comment]
Okay, so the solution here is rounding the UVs down.  Well, there may be a perf impact... although it'll be tiny on desktop.  Through isn't that common heavy on vertices anyway....

Find, VertexDecoderX86.cpp:

```c++
void VertexDecoderJitCache::Jit_TcFloatThrough() {
#ifdef _M_X64
	MOV(64, R(tempReg1), MDisp(srcReg, dec_->tcoff));
	MOV(64, MDisp(dstReg, dec_->decFmt.uvoff), R(tempReg1));
#else
	MOV(32, R(tempReg1), MDisp(srcReg, dec_->tcoff));
	MOV(32, R(tempReg2), MDisp(srcReg, dec_->tcoff + 4));
	MOV(32, MDisp(dstReg, dec_->decFmt.uvoff), R(tempReg1));
	MOV(32, MDisp(dstReg, dec_->decFmt.uvoff + 4), R(tempReg2));
#endif
}
```

To:

```c++
void VertexDecoderJitCache::Jit_TcFloatThrough() {
	MOVUPS(fpScratchReg, MDisp(srcReg, dec_->tcoff));
	ROUNDFLOORPS(fpScratchReg, R(fpScratchReg));
	MOVUPS(MDisp(dstReg, dec_->decFmt.uvoff), fpScratchReg);
}
```

This is query & dirty.  It won't work on older x86, but I think it will round it down just as a test.

If it doesn't work, alternatively you can comment out:

```c++
{&VertexDecoder::Step_TcFloatThrough, &VertexDecoderJitCache::Jit_TcFloatThrough},
```

In that file, and then in VertexDecoderCommon.cpp:

```c++
void VertexDecoder::Step_TcFloatThrough() const
{
	float *uv = (float *)(decoded_ + decFmt.uvoff);
	const float *uvdata = (const float*)(ptr_ + tcoff);
	uv[0] = (float)(int)uvdata[0];
	uv[1] = (float)(int)uvdata[1];

	gstate_c.vertBounds.minU = std::min(gstate_c.vertBounds.minU, (u16)uvdata[0]);
	gstate_c.vertBounds.maxU = std::max(gstate_c.vertBounds.maxU, (u16)uvdata[0]);
	gstate_c.vertBounds.minV = std::min(gstate_c.vertBounds.minV, (u16)uvdata[1]);
	gstate_c.vertBounds.maxV = std::max(gstate_c.vertBounds.maxV, (u16)uvdata[1]);
}
```

-[Unknown]

**Comment by hrydgard** [comment]
Through with float is quite uncommon so I wouldn't worry too much about the perf impact.

BTW, MOVUPS moves 16 bytes while we only need 8, MOVQ_xmm feels more appropriate.

here are some SSE2 floors, but maybe we can get away with something simpler:

http://dss.stephanierct.com/DevBlog/?p=8

**Comment by daniel229** [comment]
It works just change <code>void VertexDecoderJitCache::Jit_TcFloatThrough() {</code> part.

**Comment by daniel229** [comment]
PSP also gets offset,the fix maybe better to apply to HD version only?

**Comment by Gamemulatorer** [comment]
What status of this now?

**Comment by Gamemulatorer** [comment]
This is also a GPU readbacks issue?

**Comment by hrydgard** [comment]
Possibly, but there's been very little action here for a long time so not sure if it's still an issue.

**Comment by Gamemulatorer** [comment]
I tried the non hd version of this game didn't see any issue with GPU readbacks on.
![Screenshot_20230216_204512_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/219368525-cae9d680-fd64-4fb3-b054-235f1db9e846.jpg)

**Comment by Gamemulatorer** [comment]
Comparison tested in opengl only.

HD version gpu readbacks on.

https://user-images.githubusercontent.com/37603562/219369486-950959ba-37d2-4565-a7a9-cf35b9b39127.mp4

Non-hd GPU readbacks on.

https://user-images.githubusercontent.com/37603562/219370086-93eb05fc-f3c1-439e-9743-d0120873361d.mp4

Maybe the animation transition is the issue?

**Comment by unknownbrackets** [comment]
Yes, that's what the original issue is.  I still suspect that maybe the coordinates should be rounded down, but I'm not sure if that's correct or would interfere with other remasters...

-[Unknown]
